lib/api_remote.ml

```diff
@@ -113,6 +100,72 @@ let get_permalink ~(ctx : Context.t) ~(req : Slack_t.get_permalink_req) =
     ~name:(sprintf "chat.getPermalink (%s, %s)" req.channel req.message_ts)
     ~ctx `GET api_path Slack_j.read_get_permalink_res
 
+let www_form_of_get_upload_url_ext (req : Slack_t.get_upload_url_ext_req) =
+  let fields =
+    [
+      Some ("filename", req.filename);
+      Some ("length", Int.to_string req.length);
+      string_field_val req.alt_txt "alt_txt";
+      string_field_val req.snippet_type "snippet_type";
+    ]
+  in
+  list_filter_opt fields
+
+let get_upload_url_external ~(ctx : Context.t) ~(req : Slack_t.get_upload_url_ext_req) =
+  log#info "getting upload url for %s" req.filename;
+  let args = www_form_of_get_upload_url_ext req in
+  let data = Web.make_url_args args in
+  let body = `Raw ("application/x-www-form-urlencoded", data) in
+  log#info "data to upload req: %s" data;
+  request_token_auth ~ctx
+    ~name:(sprintf "files.getUploadURLExternal (%s)" req.filename)
+    ~body `POST "files.getUploadURLExternal" Slack_j.read_get_upload_url_ext_res
+
+let complete_upload_external ~(ctx : Context.t) ~(req : Slack_t.complete_upload_ext_req) =
+  log#info "completing upload url for %s" @@ Slack_j.string_of_files_v2 req.files;
+  let data = Slack_j.string_of_complete_upload_ext_req req in
+  let body = `Raw ("application/json", data) in
+  log#info "data to upload req: %s" data;
+  request_token_auth ~ctx
+    ~name:(sprintf "files.completeUploadExternal (%s)" @@ Slack_j.string_of_files_v2 req.files)
+    ~body `POST "files.completeUploadExternal" Slack_j.read_complete_upload_ext_res
+
+(** [upload_file ctx file] upload [file] to channels noted in [file.channels]
+      with content [file.content]; Not supporting file upload through form using
+      `file` currently
+      uses web API with access token *)
+let upload_file ~(ctx : Context.t) ~(file : Slack_t.files_upload_req) =
+  match file.filename, file.content with
+  | None, _ | _, None -> Exn.fail "need to supply both filename and content"
+  | Some filename, Some content ->
+    let length = String.length content in
+    let req = Slack_j.make_get_upload_url_ext_req ~filename ~length () in
+    ( match%lwt get_upload_url_external ~ctx ~req with
+    | Error e -> Lwt.return_error e
+    | Ok { upload_url; file_id; _ } ->
+      let raw_file_contents = In_channel.with_open_bin filename (fun ic -> input_all ic) in
```

Comment:
why this defaul? What if the user wants to add a snippet of a non-existing file (or a non-file)? 

You should either send the content property of the file, or read from disk if content is None.

Reply:
It's the default because I was trying to preserve the same behavior from the previous upload api. 

> What if the user wants to add a snippet of a non-existing file (or a non-file)?

I don't understand what you mean by this. My understanding is that the snippet-type is an annotation to say that the following file you want to upload should be formatted as code. It doesn't look like it's possible to add a snippet of a non-existing file? https://api.slack.com/methods/files.getUploadURLExternal#arg_snippet_type

Also it seems weird to use the file upload api and not upload a file

Reply:
> Also it seems weird to use the file upload api and not upload a file

The issue here is that they mix snippets and file uploads on the same mechanism. You can create either with these calls. _A snippet is per definition not a file_. The snippet type is meant to define a syntax highlight for the content manually. 

Maybe the source of confusion is that you can upload a snippet without a title, but this API forces you to give the file a name? (but then it tells you that [`title` is optional](https://api.slack.com/methods/files.completeUploadExternal#arg_files))

This should allow you to upload logs, for instance, or any random text content. It's one of the features we'll need it for